### PR TITLE
feat: Keras pruning example

### PR DESCRIPTION
**Pull request by Rishav1**
Addresses #354

### CODE REVIEW

File: examples/pruning/keras_integration.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    # convert class vectors to binary class matrices
+    y_train = keras.utils.to_categorical(y_train, CLASSES)
+    y_test = keras.utils.to_categorical(y_test, CLASSES)
+
+    # build model
+    model = model_fn(trial)
+
+    model.fit(x_train, y_train,
+              batch_size=BATCHSIZE,
+              callbacks=[KerasPruningCallback(trial, 'acc')],
```
Comment (toshihikoyanase) [inline examples/pruning/keras_integration.py]: Please use `val_acc` instead of `acc` because we use validation data.

File: examples/pruning/keras_integration.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                  metrics=['accuracy'])
+
+    return model
+
+
+def objective(trial):
+    # the data, split between train and test sets
+    (x_train, y_train), (x_test, y_test) = mnist.load_data()
+    x_train = x_train.reshape(60000, 784).astype('float32')/255
+    x_test = x_test.reshape(10000, 784).astype('float32')/255
```
Comment (toshihikoyanase) [inline examples/pruning/keras_integration.py]: I think it takes too much time to train models using the whole dataset. Would you use a subset of dataset like [chainer example](https://github.com/pfnet/optuna/blob/master/examples/pruning/chainer_integration.py#L28)?
We execute all examples on CircleCI every day, and we do not want to occupy the slots for a long time.

File: examples/pruning/keras_integration.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    lr = trial.suggest_uniform('lr', 1e-5, 1e-1)
+    optimizer = trial.suggest_categorical('optimizer', ['RMSprop', 'Adam', 'SGD'])
+    model.compile(loss='categorical_crossentropy',
+                  optimizer=getattr(optimizers, optimizer)(lr=lr),
+                  metrics=['accuracy'])
+
+    return model
+
+
+def objective(trial):
```
Comment (toshihikoyanase) [inline examples/pruning/keras_integration.py]: Please add `keras.backend.clear_session()` at the top of the objective function to destroy old models. The execution time will be halved by adding it. And also, it will be very helpful for users if you add a comment that explains why you call it.

File: examples/pruning/keras_integration.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                        kernel_regularizer=regularizers.l2(weight_decay)))
+    model.add(Dense(CLASSES,
+                    activation='softmax',
+                    kernel_regularizer=regularizers.l2(weight_decay)))
+
+    # Optimizer
+    lr = trial.suggest_uniform('lr', 1e-5, 1e-1)
+    optimizer = trial.suggest_categorical('optimizer', ['RMSprop', 'Adam', 'SGD'])
+    model.compile(loss='categorical_crossentropy',
+                  optimizer=getattr(optimizers, optimizer)(lr=lr),
```
Comment (toshihikoyanase) [inline examples/pruning/keras_integration.py]: I think some novice users of Optuna are not familiar with this notation, and they may have difficulty in reading this example. Could you fix the optimizer like [chainer example](https://github.com/pfnet/optuna/blob/master/examples/pruning/chainer_integration.py#L52) to simplify the example?
I think `RMSprop` is a first choice because [the official example](https://github.com/keras-team/keras/blob/master/examples/mnist_mlp.py#L46) in keras-team repository uses it.

File: examples/pruning/keras_integration.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        num_hidden = int(trial.suggest_loguniform('n_units_l{}'.format(i), 4, 128))
+        model.add(Dense(num_hidden,
+                        activation='relu',
+                        kernel_regularizer=regularizers.l2(weight_decay)))
+    model.add(Dense(CLASSES,
+                    activation='softmax',
+                    kernel_regularizer=regularizers.l2(weight_decay)))
+
+    # Optimizer
+    lr = trial.suggest_uniform('lr', 1e-5, 1e-1)
```
Comment (toshihikoyanase) [inline examples/pruning/keras_integration.py]: Please use `suggest_loguniform` instead of `suggest_uniform` to suggest `lr`. This is because the range `[1e-5, 1e-1]` seems to be log domain.

File: examples/pruning/keras_integration.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+BATCHSIZE = 128
+CLASSES = 10
+EPOCHS = 20
+
+
+def model_fn(trial):
+    # Model
+    n_layers = trial.suggest_int('n_layers', 1, 3)
+    weight_decay = trial.suggest_uniform('weight_decay', 1e-10, 1e-3)
```
Comment (toshihikoyanase) [inline examples/pruning/keras_integration.py]: [The official example](https://github.com/keras-team/keras/blob/master/examples/mnist_mlp.py#L38) in keras-team repository uses dropout instead of weight decay. I think we can optimize the dropout rate of each layer by using `trial.suggest_uniform`, and it will show the strength of Optuna's define-by-run API.
What do you think of it?

File: examples/pruning/keras_integration.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+from keras import regularizers
+
+
+BATCHSIZE = 128
+CLASSES = 10
+EPOCHS = 20
+
+
+def model_fn(trial):
+    # Model
```
Comment (toshihikoyanase) [inline examples/pruning/keras_integration.py]: Could you explain the target hyperparameters as can be seen in [chainer example](https://github.com/pfnet/optuna/blob/master/examples/pruning/chainermn_integration.py#L36)?

File: examples/pruning/keras_integration.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+from keras import optimizers
+from keras import regularizers
+
+
+BATCHSIZE = 128
+CLASSES = 10
+EPOCHS = 20
+
+
+def model_fn(trial):
```
Comment (toshihikoyanase) [inline examples/pruning/keras_integration.py]: I think this function creates an MLP model, and I prefer `create_model` in order to improve the readability.

File: examples/pruning/keras_integration.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+Keras and MNIST, where the architecture of the neural network is optimized. Throughout the
+training of neural networks, a pruner observes intermediate results and stops unpromising trials.
+
+You can run this example as follows:
+    $ python keras_integration.py
+
+"""
+
+import optuna
+from optuna.integration import KerasPruningCallback
```
Comment (toshihikoyanase) [inline examples/pruning/keras_integration.py]: Please move the import sentences of optuna after the import sentences of keras like [chainer example](https://github.com/pfnet/optuna/blob/master/examples/pruning/chainer_integration.py).

File: examples/pruning/keras_integration.py (hunk 1 above)
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Okay.

File: examples/pruning/keras_integration.py (hunk 2 above)
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Alright.

File: examples/pruning/keras_integration.py (hunk 3 above)
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Cool.

File: examples/pruning/keras_integration.py (hunk 4 above)
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Alright.

File: examples/pruning/keras_integration.py (hunk 5 above)
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Thanks for pointing that out.

File: examples/pruning/keras_integration.py (hunk 6 above)
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Alright. Dropout works better than weight_decay usually, so I guess dropout in the example would be better.

File: examples/pruning/keras_integration.py (hunk 9 above)
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Ok.

File: examples/pruning/keras_integration.py (hunk 8 above)
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Ok.

File: examples/pruning/keras_integration.py (hunk 7 above)
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: WIll do.

Comment (toshihikoyanase) [reply examples/pruning/keras_integration.py]: Thank you for your quick action.

I have nits about comments.
Could you add a period at the end of this sentence?
And also, please fix the comment style in this example like
 [chainer example](https://github.com/pfnet/optuna/blob/master/examples/pruning/chainer_integration.py#L87).
Some comments lack periods and/or start with small letters.

Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Okay. Fixed that.

Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Alright.

File: examples/pruning/keras_integration.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    $ python keras_integration.py
+
+"""
+
+import keras
+from keras.datasets import mnist
+from keras.layers import Dense
+from keras.layers import Dropout
+from keras.models import Sequential
+
```
[the remark this answers is not in the document]
Comment (Rishav1) [reply examples/pruning/keras_integration.py]: Alright. I'll do that.

**Comment by g-votte** [comment]
Resolved the conflict.

**Comment by g-votte** [comment]
Merged. Thanks for your contribution!
